### DISCUSSION TITLE
Build hangs indefinitely because of telemetry API

### DISCUSSION THREAD

**Discussion by tsoos99dev** [vercel/next.js#71063] [Help]
### Link to the code that reproduces this issue

https://github.com/tsoos99dev/reproduction-app

### To Reproduce

1. Start the build with any setup, it doesn't matter.
2. Wait till the console shows ○  (Static)  prerendered as static content

### Current vs. Expected behavior

Following the steps results in the build getting stuck. I was expecting the build to finish.

So the problem seems to be with telemetry. I have no idea how anyone else can reproduce this. 

This line hangs at the end of the build in `next/dist/build/index.js`:
`await nextBuildSpan.traceChild("telemetry-flush").traceAsyncFn(()=>telemetry.flush());`
right after printing the `TreeView`.

This means the queue in `next/dist/telemetry/storage.js` isn't empty. More specifically we're waiting on this line inside the promises left in the queue:
```
const res = (0, _postpayload._postPayload)(`https://telemetry.nextjs.org/api/v1/record`, {
                context,
                meta,
                events: events.map(({ eventName, payload })=>({
                        eventName,
                        fields: payload
                    }))
            }, postController.signal);
```
which is connected to the promise by the call to `submitRecord` inside `record` which is also in `next/dist/build/telemetry/storage.js`.
I've added a manual 5 second timeout to the `_postPayload` call and that fixed the issue. I'm seeing 4 promises rejected at the end. They always seem to be the same:
```
Aborted:  [
  {
    eventName: 'NEXT_BUILD_COMPLETED',
    payload: {
      durationInSeconds: 0,
      totalAppPagesCount: 3,
      totalPageCount: 0,
      hasDunderPages: false,
      hasTestPages: false
    }
  }
]
Aborted:  [
  {
    eventName: 'NEXT_LINT_CHECK_COMPLETED',
    payload: {
      durationInSeconds: 0,
      eslintVersion: '8.57.1',
      lintedFilesCount: 2,
      lintFix: false,
      nextEslintPluginVersion: '14.2.13',
      nextEslintPluginErrorsCount: 0,
      nextEslintPluginWarningsCount: 0,
      nextRulesEnabled: [Object],
      buildLint: true
    }
  }
]
Aborted:  [
  {
    eventName: 'NEXT_TYPE_CHECK_COMPLETED',
    payload: {
      durationInSeconds: 0,
      typescriptVersion: '5.6.2',
      inputFilesCount: 6,
      totalFilesCount: 428,
      incremental: true
    }
  }
]
Aborted:  [
  {
    eventName: 'NEXT_BUILD_FEATURE_USAGE',
    payload: { featureName: 'experimental/optimizeCss', invocationCount: 0 }
  },
  {
    eventName: 'NEXT_BUILD_FEATURE_USAGE',
    payload: {
      featureName: 'experimental/nextScriptWorkers',
      invocationCount: 0
    }
  },
  {
    eventName: 'NEXT_BUILD_FEATURE_USAGE',
    payload: { featureName: 'optimizeFonts', invocationCount: 1 }
  },
  {
    eventName: 'NEXT_BUILD_FEATURE_USAGE',
    payload: { featureName: 'experimental/ppr', invocationCount: 0 }
  }
]
```

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #45~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Sep 11 15:25:05 UTC 2
  Available memory (MB): 15730
  Available CPU cores: 16
Binaries:
  Node: 20.17.0
  npm: 10.8.3
  Yarn: 1.22.22
  pnpm: N/A
Relevant Packages:
  next: 14.2.14 // Latest available version is detected (14.2.14).
  eslint-config-next: 14.2.13
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

I tested against 15.0.0-canary.177 and I got the same result. The issue had been introduced before 14.2.14 and hasn't been fixed so far. Not sure when it was introduced exactly.

**Comment by samcx** [comment]
@tsoos99dev Unfortunately, not able to replicate this. If you're able to figure out how this is affecting you, please let us know!

**Reply by tsoos99dev** [reply]
Fair enough. A discussion is more suitable at this point. So I collected all the telemetry call payloads and wrote a separate python script to send them to the telemetry API endpoint. Same happens, the POST calls just hang, not all of them though. 
[telemetry.txt](https://github.com/user-attachments/files/17329109/records.txt)

I figured out that it has to do with the content length. Only long enough payloads hang. From the list of payloads I attached only the 2., 5., and the 9. fail. Their content lengths are:
- 1966 bytes
- 1810 bytes
- 4837 bytes
Then I wrote a small python script that generates random ascii characters, and sends them in a simple json object, to figure out the content length limit. The test payload looks like this:
```json
{
    "data": "asdfno934h^%$#$R@#$gsdf..."
}
```

I found that at exactly 1423 bytes the call hangs and it's repeatable, but only on my homelab. The same call with much larger content lengths works fine on my macbook which is on the same network. If I send the call from my homelab to another external API I get a response back just fine. 

Here are the request headers I'm sending:
```
Host: telemetry.nextjs.org
Accept: */*
Accept-Encoding: gzip, deflate
User-Agent: Python/3.10 aiohttp/3.10.9
Content-Length: 1422
Content-Type: application/json
```

Although I'd like to figure out why the call is failing my suggestion would be to introduce a timeout in the telemetry module and print out a warning instead of letting the build hang forever.
